cpp/oneapi/dal/backend/primitives/reduction/reduction_dpc.cpp

```diff
@@ -0,0 +1,141 @@
+*******************************************************************************/
+
+#include <type_traits>
+
+#include "oneapi/dal/backend/primitives/reduction/reduction_dpc.hpp"
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_rw_dpc.hpp"
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_cw_dpc.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+inline sycl::event reduce_rm_rw(sycl::queue& q,
+                                const ndview<Float, 2, ndorder::c>& input,
+                                ndview<Float, 1>& output,
+                                const BinaryOp& binary,
+                                const UnaryOp& unary,
+                                const event_vector& deps) {
+    using kernel_t = reduction_rm_rw<Float, BinaryOp, UnaryOp>;
+    const auto width = input.get_dimension(1);
+    const auto height = input.get_dimension(0);
+    const auto stride = input.get_leading_stride();
+    const auto* inp_ptr = input.get_data();
+    auto* out_ptr = output.get_mutable_data();
+    const kernel_t kernel(q);
+    return kernel(inp_ptr, out_ptr, width, height, stride, binary, unary, deps);
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+inline sycl::event reduce_rm_cw(sycl::queue& q,
+                                const ndview<Float, 2, ndorder::c>& input,
+                                ndview<Float, 1>& output,
+                                const BinaryOp& binary,
+                                const UnaryOp& unary,
+                                const event_vector& deps) {
+    using kernel_t = reduction_rm_cw<Float, BinaryOp, UnaryOp>;
+    const auto width = input.get_dimension(1);
+    const auto height = input.get_dimension(0);
+    const auto stride = input.get_leading_stride();
+    const auto* inp_ptr = input.get_data();
+    auto* out_ptr = output.get_mutable_data();
+    const kernel_t kernel(q);
+    return kernel(inp_ptr, out_ptr, width, height, stride, binary, unary, deps);
+}
+
+template <typename Float, ndorder order, typename BinaryOp, typename UnaryOp>
+sycl::event reduce_rows(sycl::queue& q,
+                        const ndview<Float, 2, order>& input,
+                        ndview<Float, 1>& output,
+                        const BinaryOp& binary,
+                        const UnaryOp& unary,
+                        const event_vector& deps) {
+    ONEDAL_ASSERT(input.get_dimension(0) <= output.get_dimension(0));
+    if constexpr (order == ndorder::c) {
+        return reduce_rm_rw(q, input, output, binary, unary, deps);
+    }
+    else {
+        auto input_tr = input.t();
+        return reduce_rm_cw(q, input_tr, output, binary, unary, deps);
+    }
+    ONEDAL_ASSERT(false);
+    return sycl::event{};
+}
+
+template <typename Float, ndorder order, typename BinaryOp, typename UnaryOp>
+sycl::event reduce_cols(sycl::queue& q,
```

Comment:
Is this reduces columns so output contains just 1 column?

Reply:
It reduces columns. So the result will be just one "mean" sample